### PR TITLE
[RISCV][GlobalISel] Select G_GLOBAL_VALUE

### PR DESCRIPTION
**Pull request by michaelmaitland**
G_GLOBAL_VALUE should be lowered into an absolute address if `-codemodel=small` is used or into a PC-relative if `-codemodel=medium` is used.

PR #68380 tried to create special instructions to do this, but I don't see why we need to do that.

### CODE REVIEW

**Comment by jrtc27** [comment]
PIC/PIE?

**Comment by jrtc27** [comment]
> PR https://github.com/llvm/llvm-project/pull/68380 tried to create special instructions to do this, but I don't see why we need to do that.

Presumably "because SDAG does". Though I still to this day don't understand the point of most of those SDAG indirections; we used to just create the machine nodes directly. Seemed to just add a bunch of busywork jumping between different equivalent forms.

**Comment by michaelmaitland** [comment]
> PIC/PIE?

Does SDAG do something special for PIC/PIE? What did you have in mind?

**Comment by topperc** [comment]
> > PR #68380 tried to create special instructions to do this, but I don't see why we need to do that.
> 
> Presumably "because SDAG does". Though I still to this day don't understand the point of most of those SDAG indirections; we used to just create the machine nodes directly. Seemed to just add a bunch of busywork jumping between different equivalent forms.

The only that was probably actually needed was RISCVISD::ADD_LO for RISCVDAGToDAGISel::SelectAddrRegImm.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 1]
```diff
@@ -483,6 +489,66 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
...
+         "Expected G_GLOBAL_VALUE");
+  auto GV = MI.getOperand(1).getGlobal();
+  if (GV->isThreadLocal()) {
+    // TODO: implement this case.
+    return false;
+  }
+
+  switch (TM.getCodeModel()) {
+  default:
+    report_fatal_error("Unsupported code model for lowering");
```
Comment (tschuett) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:504]:
```
reportGISelFailure(const_cast<MachineFunction &>(MF), *TPC, *MORE,
                       "riskv-selector", "unknown code model",
                       MI);
```
turns the error into a remark.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 2]
```diff
@@ -483,6 +489,66 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
...
+      return false;
+
+    MI.eraseFromParent();
+    return true;
+  }
+  case CodeModel::Medium: {
+    // Must lie within *any* 2GiB range. The generates (addi (auipc
+    // %pcrel_hi(sym)) %pcrel_lo(auipc)).
+    Register AddrHiDest = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    MachineInstr *AddrHi = MIB.buildInstr(RISCV::AUIPC)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:531]: This needs to be emitted as PseudoLGA and get expanded later. The pcrel_lo relocation needs to reference a label that points to the AUIPC instruction itself, not the global. We can't do that from inside the instruction selector.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp (hunk 1 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp]: 64daec66e4d03d0a002d3973bc25dd7cedf312e8

**Comment by jrtc27** [comment]
isPositionIndependent() needs to use LLA/LGA as appropriate based on whether the symbol is local. Plus -mcmodel=medium needs to use LGA for extern weak. See RISCVTargetLowering::getAddr for the full set of behaviours needed.

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/global-value.mir [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  bb.1.entry:
+    ; RV64-SMALL-LABEL: name: global_addr
+    ; RV64-SMALL: [[LUI:%[0-9]+]]:gpr = LUI @x + 4
+    ; RV64-SMALL-NEXT: [[ADDI:%[0-9]+]]:gpr = ADDI [[LUI]], target-flags(riscv-lo) @x
+    ; RV64-SMALL-NEXT: $x10 = COPY [[ADDI]]
+    ; RV64-SMALL-NEXT: PseudoRET implicit $x10
+    ;
+    ; RV64-MED-LABEL: name: global_addr
+    ; RV64-MED: [[PseudoLLA:%[0-9]+]]:gpr = PseudoLLA @x
+    ; RV64-MED-NEXT: $x10 = COPY [[PseudoLLA]]
```
Comment (michaelmaitland) [inline llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/global-value.mir:36]: This isn't quite right. I need to take another look.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 4]
```diff
@@ -483,6 +489,78 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
...
+    // Emit LGA/LLA instead of the sequence it expands to because the pcrel_lo
+    // relocation needs to reference a label that points to the auipc
+    // instruction itself, not the global. This cannot be done inside the
+    // instruction selector.
+    if (GV->hasExternalWeakLinkage())
+      // An extern weak symbol may be undefined, i.e. have value 0, which may
+      // not be within 2GiB of PC, so use GOT-indirect addressing to access the
+      // symbol. This generates the pattern (PseudoLGA sym), which expands to
+      // (ld (addi (auipc %got_pcrel_hi(sym)) %pcrel_lo(auipc))).
+      Result = MIB.buildInstr(RISCV::PseudoLGA)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:543]: Need to add a MemOperand for the GOT.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 5]
```diff
@@ -483,6 +489,78 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
...
+  }
+
+  switch (TM.getCodeModel()) {
+  default: {
+    reportGISelFailure(const_cast<MachineFunction &>(*MF), *TPC, *MORE,
+                       getName(), "Unsupported code model for lowering", MI);
+    return false;
+  }
+  case CodeModel::Small: {
+    // Must lie within a single 2 GiB address range and must lie between
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:509]: Still missing the `isPositionIndependent()` from the top of SelectionDAG's `getAddr` that uses LGA/LLA even for the small code model.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp (hunk 4 above)
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:543]: This isn't tested?

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 6]
```diff
@@ -483,6 +489,126 @@ bool RISCVInstructionSelector::selectConstant(MachineInstr &MI,
...
+  case CodeModel::Small: {
+    Register DefReg = MI.getOperand(0).getReg();
+    const LLT DefTy = MRI.getType(DefReg);
+    MachineInstr *Result = nullptr;
+
+    // When HWASAN is used and tagging of global variables is enabled
+    // they should be accessed via the GOT, since the tagged address of a global
+    // is incompatible with existing code models. This also applies to non-pic
+    // mode.
+    if (TM.isPositionIndependent() && GV->isDSOLocal() &&
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:518]: Should this be outside the switch like it is in getAddr?

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:518]: I meant the whole position independent code block.  It's independent of code model.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 7]
```diff
@@ -506,50 +507,97 @@ bool RISCVInstructionSelector::selectGlobalValue(
...
+                   .addGlobalAddress(GV, 0);
+    } else if (Subtarget->allowTaggedGlobals() ||
+               (TM.isPositionIndependent() && !GV->isDSOLocal())) {
+      // Use PC-relative addressing to access the GOT for this symbol, then
+      // load the address from the GOT. This generates the pattern (PseudoLGA
+      // sym), which expands to (ld (addi (auipc %got_pcrel_hi(sym))
+      // %pcrel_lo(auipc))).
+      MachineFunction &MF = *MI.getParent()->getParent();
+      MachineMemOperand *MemOp =
+          MI.getParent()->getParent()->getMachineMemOperand(
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:526]: Reuse MF instead of all the getParents

**Review by topperc — CHANGES_REQUESTED** [review]
Missing rv32 test

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**